lib/xnvme_crc.c

```diff
@@ -0,0 +1,587 @@
+	0xd53368f1dcdf0249ULL, 0xaa5d9839ecea8b30ULL, 0x449080a64ce15841ULL,
+	0x3bfe706e7cd4d138ULL, 0xba4d61362c8a4ab3ULL, 0xc52391fe1cbfc3caULL,
+	0x8df265d5d4a0eeceULL, 0xf29c951de49567b7ULL, 0x732f8445b4cbfc3cULL,
+	0x0c41748d84fe7545ULL, 0x6bad6d5ebd3716b7ULL, 0x14c39d968d029fceULL,
+	0x95708ccedd5c0445ULL, 0xea1e7c06ed698d3cULL, 0xa2cf882d2576a038ULL,
+	0xdda178e515432941ULL, 0x5c1269bd451db2caULL, 0x237c997575283bb3ULL,
+	0xcdb181ead523e8c2ULL, 0xb2df7122e51661bbULL, 0x336c607ab548fa30ULL,
+	0x4c0290b2857d7349ULL, 0x04d364994d625e4dULL, 0x7bbd94517d57d734ULL,
+	0xfa0e85092d094cbfULL, 0x856075c11d3cc5c6ULL, 0x134d926535897936ULL,
+	0x6c2362ad05bcf04fULL, 0xed9073f555e26bc4ULL, 0x92fe833d65d7e2bdULL,
+	0xda2f7716adc8cfb9ULL, 0xa54187de9dfd46c0ULL, 0x24f29686cda3dd4bULL,
+	0x5b9c664efd965432ULL, 0xb5517ed15d9d8743ULL, 0xca3f8e196da80e3aULL,
+	0x4b8c9f413df695b1ULL, 0x34e26f890dc31cc8ULL, 0x7c339ba2c5dc31ccULL,
+	0x035d6b6af5e9b8b5ULL, 0x82ee7a32a5b7233eULL, 0xfd808afa9582aa47ULL,
+	0x4d364994d625e4daULL, 0x3258b95ce6106da3ULL, 0xb3eba804b64ef628ULL,
+	0xcc8558cc867b7f51ULL, 0x8454ace74e645255ULL, 0xfb3a5c2f7e51db2cULL,
+	0x7a894d772e0f40a7ULL, 0x05e7bdbf1e3ac9deULL, 0xeb2aa520be311aafULL,
+	0x944455e88e0493d6ULL, 0x15f744b0de5a085dULL, 0x6a99b478ee6f8124ULL,
+	0x224840532670ac20ULL, 0x5d26b09b16452559ULL, 0xdc95a1c3461bbed2ULL,
+	0xa3fb510b762e37abULL, 0x35d6b6af5e9b8b5bULL, 0x4ab846676eae0222ULL,
+	0xcb0b573f3ef099a9ULL, 0xb465a7f70ec510d0ULL, 0xfcb453dcc6da3dd4ULL,
+	0x83daa314f6efb4adULL, 0x0269b24ca6b12f26ULL, 0x7d0742849684a65fULL,
+	0x93ca5a1b368f752eULL, 0xeca4aad306bafc57ULL, 0x6d17bb8b56e467dcULL,
+	0x12794b4366d1eea5ULL, 0x5aa8bf68aecec3a1ULL, 0x25c64fa09efb4ad8ULL,
+	0xa4755ef8cea5d153ULL, 0xdb1bae30fe90582aULL, 0xbcf7b7e3c7593bd8ULL,
+	0xc399472bf76cb2a1ULL, 0x422a5673a732292aULL, 0x3d44a6bb9707a053ULL,
+	0x759552905f188d57ULL, 0x0afba2586f2d042eULL, 0x8b48b3003f739fa5ULL,
+	0xf42643c80f4616dcULL, 0x1aeb5b57af4dc5adULL, 0x6585ab9f9f784cd4ULL,
+	0xe436bac7cf26d75fULL, 0x9b584a0fff135e26ULL, 0xd389be24370c7322ULL,
+	0xace74eec0739fa5bULL, 0x2d545fb4576761d0ULL, 0x523aaf7c6752e8a9ULL,
+	0xc41748d84fe75459ULL, 0xbb79b8107fd2dd20ULL, 0x3acaa9482f8c46abULL,
+	0x45a459801fb9cfd2ULL, 0x0d75adabd7a6e2d6ULL, 0x721b5d63e7936bafULL,
+	0xf3a84c3bb7cdf024ULL, 0x8cc6bcf387f8795dULL, 0x620ba46c27f3aa2cULL,
+	0x1d6554a417c62355ULL, 0x9cd645fc4798b8deULL, 0xe3b8b53477ad31a7ULL,
+	0xab69411fbfb21ca3ULL, 0xd407b1d78f8795daULL, 0x55b4a08fdfd90e51ULL,
+	0x2ada5047efec8728ULL};
+
+static inline uint64_t
+crc64_rocksoft_refl_base(uint64_t seed, const uint8_t *buf, uint64_t len)
+{
+	uint64_t i, crc = ~seed;
+
+	for (i = 0; i < len; i++) {
+		uint8_t byte = buf[i];
+		crc = crc64_rocksoft_refl_table[(uint8_t)crc ^ byte] ^ (crc >> 8);
+	}
+
+	return ~crc;
+}
+
+uint64_t
+xnvme_crc64_nvme(const void *buf, size_t len, uint64_t crc)
+{
+	return crc64_rocksoft_refl_base(crc, (const uint8_t *)buf, len);
```

Comment:
**Nitpick**. The CRC64 used in NVM Express is NOT called "Rocksoft". It's just called the "NVM Express 64b CRC" (it uses an NVMe unique polynomial). I know the above is because isa-l calls it that, but could do a `#define` here to point out that all the other implementations misunderstood something ;)

Reply:
Yeah correct, Keith also had the same comment in fio patches.